Compatibility layer additions for cisco/libest

julek-wolfssl opened the pull request.

julek-wolfssl committed:
Add `--enable-libest` option to configure.ac

Refactoring and adding defines for functions

julek-wolfssl committed:
Implement ASN1_get_object

julek-wolfssl committed:
Implement/stub the following:

- `NID_pkcs9_challengePassword` - added
- `wolfSSL_OPENSSL_cleanse` - implemented
- `wolfSSL_X509_REQ_add1_attr_by_NID` - stubbed
- `wolfSSL_c2i_ASN1_OBJECT` - stubbed

julek-wolfssl committed:
Implement/stub the following functions:

- X509_REQ_sign_ctx
- X509_REQ_get_subject_name
- X509_REQ_set_version
- X509_NAME_print_ex_fp
- X509_STORE_CTX_get0_parent_ctx
- wolfSSL_PKCS7_encode_certs

Add cms.h file to avoid including the OpenSSL version.

julek-wolfssl committed:
Implement wolfSSL_PKCS7_to_stack and wolfSSL_d2i_ASN1_OBJECT

- I also implemented wolfSSL_c2i_ASN1_OBJECT which was previously a stub.
- More configure.ac flags added to libest option

julek-wolfssl committed:
Implement/stub the following:

- X509_get0_extensions
- X509_to_X509_REQ
- i2d_X509_REQ_bio
- X509v3_get_ext_count
- i2d_PKCS7_bio
Additional changes:
- Added a wc_PKCS7_VerifySignedData call to wolfSSL_d2i_PKCS7_bio to populate the PKCS7 struct with parsed values
- wc_PKCS7_VerifySignedData_ex -> wc_PKCS7_VerifySignedData

julek-wolfssl committed:
Implement/stub:

- wc_DhKeyCopy
- SSL_CTX_set_srp_strength
- SSL_get_srp_username
- X509_REQ_get_attr_by_NID
- X509_REQ_get_attr
- X509_ATTRIBUTE
- wolfSSL_DH_dup
Add srp.h file with SRP_MINIMAL_N

julek-wolfssl committed:
Implement/stub:

- X509_REQ_print_fp
- X509_print_fp
- DHparams_dup

julek-wolfssl committed:
PKCS7 changes

- Allow PKCS7_EncodeSigned to be called with a zero content length
- wc_HashUpdate now doesn't error out on zero length data
- First cert in wolfSSL_PKCS7_encode_certs is treated as main cert and the PKCS7 struct is initialized with it
- wolfSSL_BIO_get_mem_data returns the buffer from the last bio in chain

julek-wolfssl committed:
WIP

- wolfSSL_BIO_ctrl_pending ignore BASE64 bio's as well now
- Save the last Finished messages sent or received in the WOLFSSL struct
- Implement wolfSSL_CTX_set_max_proto_version
- wolfSSL_d2i_X509_bio now uses wolfSSL_BIO_read so that the entire chain is properly read from the BIO

julek-wolfssl committed:
Add CSR parsing capabilities to ParseCertRelative and wc_GetPubX509

- wolfSSL_BIO_get_mem_data now returns the last memory BIO in the chain
- Change wolfSSL_BIO_pending calls to wolfSSL_BIO_get_len calls to get accurate length depending on BIO
- Refactor X509 and X509_REQ functions to reuse similar code
- X509 and X509_REQ i2d functions now generate their DER outputs instead of returning the input DER
- Signature generated by wolfSSL_X509_resign_cert is now saved in the x509->sig buffer and added when calling *i2d
- Add test_wolfSSL_d2i_X509_REQ

julek-wolfssl committed:
Implement wolfSSL_X509_REQ_verify

julek-wolfssl committed:
Implement OpenSSL Compat API:

- Implement lhash as a stack with hash members
- wolfSSL_lh_retrieve
- wolfSSL_LH_strhash
- IMPLEMENT_LHASH_COMP_FN
- IMPLEMENT_LHASH_HASH_FN
- wolfSSL_sk_CONF_VALUE_new
- wolfSSL_sk_CONF_VALUE_free
- wolfSSL_sk_CONF_VALUE_num
- wolfSSL_sk_CONF_VALUE_value
- wolfSSL_NCONF_new
- wolfSSL_NCONF_get_string
- wolfSSL_NCONF_get_section
- wolfSSL_lh_WOLFSSL_CONF_VALUE_retrieve
- wolfSSL_CONF_modules_load

julek-wolfssl committed:
OpenSSL Compat layer

Implement WOLFSSL_CONF_VALUE:
- wolfSSL_CONF_VALUE_new
- wolfSSL_CONF_VALUE_new_values
- wolfSSL_CONF_add_string
- wolfSSL_X509V3_conf_free
- wolfSSL_sk_CONF_VALUE_push
- wolfSSL_NCONF_load
- wolfSSL_NCONF_free
- wolfSSL_CONF_new_section
- wolfSSL_CONF_get_section
Implment some buffer functions
- wolfSSL_strlcat
- wolfSSL_strlcpy

julek-wolfssl committed:
OpenSSL Compat layer

Implement/stub:
- wolfSSL_NCONF_get_number
- wolfSSL_EVP_PKEY_CTX_ctrl_str
- wolfSSL_PKCS12_verify_mac
- wc_PKCS12_verify_ex
- wolfSSL_BIO_new_fd
- wolfSSL_X509_sign_ctx
- wolfSSL_ASN1_STRING_cmp
- wolfSSL_ASN1_TIME_set_string
- X509V3_EXT_add_nconf
- X509V3_set_nconf
Implement TXT_DB functionality:
- wolfSSL_TXT_DB_read
- wolfSSL_TXT_DB_free
- wolfSSL_TXT_DB_create_index
- wolfSSL_TXT_DB_get_by_index

julek-wolfssl committed:
OpenSSL Compat layer

Implment/stub:
- wolfSSL_X509_NAME_delete_entry
- wolfSSL_X509_get_ext_by_OBJ
- wolfSSL_a2i_ASN1_INTEGER
- X509V3_parse_list
- wolfSSL_TXT_DB_write
- wolfSSL_TXT_DB_insert
- wolfSSL_EVP_PKEY_get_default_digest_nid

julek-wolfssl committed:
OpenSSL Compat layer

Implement/stub:
- wolfSSL_X509V3_EXT_add_nconf
- wolfSSL_EVP_PKEY_copy_parameters

julek-wolfssl committed:
Pass libest estclient_simple example

julek-wolfssl committed:
WIP

julek-wolfssl committed:
Don't attempt TLS 1.3  if server options disable it

julek-wolfssl committed:
Tests starting to pass

julek-wolfssl committed:
Set default digest NID

julek-wolfssl committed:
wolfSSL_X509_resign_cert updates x509 der buffer as well

julek-wolfssl committed:
Add CRL loading to wolfSSL_PEM_X509_INFO_read_bio

julek-wolfssl committed:
Add DSA support to ConfirmSignature and add DSAwithSHA256

julek-wolfssl committed:
Fixes

- Fix challengePw copy in ReqCertFromX509
- Proper header length in wolfSSL_PEM_X509_X509_CRL_X509_PKEY_read_bio
- Special case for extended key usage in wolfSSL_OBJ_cmp
- Numerical input in wolfSSL_OBJ_txt2obj can just be encoded with EncodePolicyOID. Searching for the sum can return wrong values since they are not unique.

julek-wolfssl committed:
Init wolfSSL_X509_REQ_add_extensions

julek-wolfssl committed:
Extension manipulation

julek-wolfssl committed:
Implement wolfSSL_BIO_*_connect and wolfSSL_BIO_set_conn_port

Forgot to commit csr.dsa.pem for api.c

julek-wolfssl committed:
Protect against invalid write in RsaPad_PSS

julek-wolfssl committed:
Dynamically allocate memory in wolfSSL_i2d_PKCS7_bio

julek-wolfssl committed:
Store more certs in PKCS7 struct

julek-wolfssl committed:
wolfSSL_X509_REQ_add1_attr_by_txt for libest

julek-wolfssl committed:
Support more extensions

julek-wolfssl committed:
Implement wolfSSL_X509_check_private_key

julek-wolfssl committed:
Fix max SSL version handling for client

Enable CRL when adding one to store

julek-wolfssl committed:
Add DSA support to x509 certs

julek-wolfssl committed:
Differentiate between server and client sessions

This is important is the client and server share memory space. If a server and client both save the same session in SessionCache it may cause inconsistencies. The hash of the sessionID will be the same causing one of the sides to overwrite the other. A possible problem is that the peer certificate will be incorrect for one of the sides.

julek-wolfssl committed:
check null

julek-wolfssl committed:
Internal unit tests

julek-wolfssl committed:
Plug memory leaks

julek-wolfssl committed:
Rebase and test fixes

julek-wolfssl committed:
Jenkins fixes

toddouska commented:
Jacob please do initial review and then pass to me.

JacobBarthelmeh commented on wolfssl/wolfcrypt/types.h:
```diff
@@ -647,10 +648,11 @@ decouple library dependencies with standard string, memory and so on.
             #define XTOUPPER(c)     toupper((c))
-            #define XISALPHA(c)     isalpha((c))
         #endif
+        #define XISALNUM(c)     isalnum((c))
```
Can we limit when this is compiled. Maybe only with OPENSSL_EXTRA. That might help later on with porting wolfSSL.

JacobBarthelmeh commented on src/ssl.c:
```diff
@@ -40229,6 +42474,37 @@ void wolfSSL_BUF_MEM_free(WOLFSSL_BUF_MEM* buf)
+size_t wolfSSL_strlcpy(char *dst, const char *src, size_t dstSize)
+{
+    size_t i;
+
```
null checks on dst and src here

JacobBarthelmeh reviewed (changes requested):
Compile time issue of:

```
./configure --enable-libest && make

In file included from ./wolfssl/openssl/crypto.h:30:
./wolfssl/openssl/conf.h:60:54: error: unknown type name 'wolf_sk_compare_cb'
WOLFSSL_API WOLFSSL_STACK *wolfSSL_sk_CONF_VALUE_new(wolf_sk_compare_cb ...
                                                     ^
1 error generated.
make[2]: *** [wolfcrypt/src/src_libwolfssl_la-asn.lo] Error 1
```

julek-wolfssl commented on src/internal.c:
```diff
@@ -15488,7 +15568,11 @@ int ProcessReply(WOLFSSL* ssl)
-                    else if (!IsAtLeastTLSv1_3(ssl->version)) {
+                    else if (!IsAtLeastTLSv1_3(ssl->version)
+#if defined(WOLFSSL_TLS13) && !defined(WOLFSSL_NO_TLS12)
+                            || !TLSv1_3_Capable(ssl)
```
Previously wolfSSL would ignore the options set by the user if TLS 1.3 was compiled in and the other side supports TLS 1.3 and wolfTLS_client_method or wolfSSLv23_client_method was used to create a SSL_CTX object. This change checks that this is permitted by the additional options set by the user.

JacobBarthelmeh commented:
The changes in the PR look good. I went to compile libest with it and ran into a couple errors on the libest side (this all looks like issues that the can be resolved in the libest fork). Was using the wolfssl branch here https://github.com/julek-wolfssl/libest/tree/wolfssl when testing linking to the port. wolfSSL was compiled using "./configure --enable-libest" and the libest fork was compiled using "./autogen.sh && ./configure --enable-wolfssl && make".

julek-wolfssl commented:
If you are talking about the "specified bound <num> equals destination size" warnings then these are issues in the libest library. I think that fixing these is outside the scope of this port.

JacobBarthelmeh commented:
Sent the list by email for review. It could be that you have some local changes yet pushed? I checked though that the API and macros it was listing as not found where indeed in this PR though. So has more to do with headers or configuration in the libest changes that I was testing with I think.

SparkiDev commented on wolfssl/wolfcrypt/pkcs7.h:
```diff
@@ -48,7 +48,7 @@
 /* Max number of certificates that PKCS7 structure can parse */
 #ifndef MAX_PKCS7_CERTS
-    #define MAX_PKCS7_CERTS 4
+    #define MAX_PKCS7_CERTS 15
```
Should this be low by default and made larger for specific builds?

toddouska replied:
Agreed, that looks too big for a default.

SparkiDev commented on wolfssl/wolfcrypt/wc_port.h:
```diff
@@ -593,6 +594,9 @@ WOLFSSL_API int wolfCrypt_Cleanup(void);
         #include <dirent.h>
         #include <unistd.h>
         #include <sys/stat.h>
+        #define XWRITE      write
```
Not available on all platforms.
Is the BIO code going to compile on Windows?

julek-wolfssl replied:
`XWRITE` is excluded from Windows builds. I just built wolfSSL with libest support on Windows and successfully ran the testsuite tests.

SparkiDev commented:
Added a couple of comments for Todd to look over. Didn't do a complete review.

toddouska commented on src/internal.c:
```diff
@@ -5255,6 +5259,44 @@ int SetSSL_CTX(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
+    if (ssl->version.minor == TLSv1_3_MINOR &&
+     (ssl->options.mask & SSL_OP_NO_TLSv1_3) == SSL_OP_NO_TLSv1_3) {
+        WOLFSSL_MSG("\tOption set to not allow TLSv1.3, Downgrading");
+        ssl->version.minor = TLSv1_2_MINOR;
```
And only downgrade if downgrade is on and downgrade version is allowed?

julek-wolfssl replied:
The reason why this shouldn't be a hard error is that if a CTX object is created using `wolfTLS_client_method_ex` then this will return the highest supported protocol version. libest then sets the maximum protocol version using `SSL_CTX_set_max_proto_version`. [This behaviour can be seen used here.](https://github.com/julek-wolfssl/libest/blob/74e1d0bb0f5a64d9be8b9d1dca1dc448cf5aab16/src/est/est_client.c#L810)

I added checking if downgrading is allowed.

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -2022,6 +2022,8 @@ struct WOLFSSL_CERT_MANAGER {
 #if defined(HAVE_ECC) || defined(HAVE_ED25519) || defined(HAVE_ED448)
     short           minEccKeySz;         /* minimum allowed ECC key size */
 #endif
+    wolfSSL_Mutex   refMutex;   /* reference count mutex */
```
This was implemented because there was a double free bug when libest called `SSL_CTX_set_cert_store` but I can't replicate it now with the old implementation. I can get rid of the reference counter for this struct if you would like. This is used like other reference counters (init in `wolfSSL_CertManagerNew`, increment in `wolfSSL_CertManager_up_ref`, decrement in `wolfSSL_CertManagerFree`).

julek-wolfssl commented on wolfssl/wolfcrypt/types.h:
```diff
@@ -495,11 +495,12 @@ decouple library dependencies with standard string, memory and so on.
         #define XSTRSTR(s1,s2)    strstr((s1),(s2))
         #define XSTRNSTR(s1,s2,n) mystrnstr((s1),(s2),(n))
         #define XSTRNCMP(s1,s2,n) strncmp((s1),(s2),(n))
+        #define XSTRCMP(s1,s2)    strcmp((s1),(s2))
```
This is used in the `WOLFSSL_CONF` part of the code where we are comparing two C strings. It would be the same as calling `XSTRLEN` twice, checking length, and comparing with `XSTRNCMP`. Using `XSTRCMP` makes the code cleaner and has the same result. I can refactor to avoid `XSTRCMP`.

julek-wolfssl committed:
Check if downgrading is allowed in SetSSL_CTX

Pkcs7 cert limit based on build

SparkiDev commented on src/ssl.c:
```diff
@@ -38186,22 +40151,20 @@ void* wolfSSL_GetDhAgreeCtx(WOLFSSL* ssl)
-
-    WOLFSSL_X509 *wolfSSL_PEM_read_bio_X509(WOLFSSL_BIO *bp, WOLFSSL_X509 **x,
-                                                 pem_password_cb *cb, void *u)
+    static WOLFSSL_X509 *wolfSSL_PEM_read_bio_X509_X509_REQ(WOLFSSL_BIO *bp,
```
Rename

SparkiDev commented on src/ssl.c:
```diff
@@ -40129,17 +42400,39 @@ void* wolfSSL_GetDhAgreeCtx(WOLFSSL* ssl)
+            return WOLFSSL_FAILURE;
+        }
+
+        der = (byte*)wolfSSL_X509_get_der(x509, &derSz);
```
Can't we get the public key out of the WOLFSSL_X509 structure?

julek-wolfssl replied:
We probably could but we don't have any API that takes in just a public key without a certificate for verification.

SparkiDev replied:
The function is checking that the public key matches the private key.
OpenSSL doesn't validate the certificate, it just extracts the public key and compares to the private key.

julek-wolfssl replied:
Refactored wc_CheckPrivateKey to accept buffers and added wc_CheckPrivateKeyCert if we want to pass in DecodedCert instead.

SparkiDev commented on src/ssl.c:
```diff
@@ -48969,16 +51699,271 @@ int wolfSSL_X509_REQ_sign(WOLFSSL_X509 *req, WOLFSSL_EVP_PKEY *pkey,
+        return WOLFSSL_FAILURE;
+}
+
+static int wolfSSL_regen_X509_REQ_der_buffer(WOLFSSL_X509* x509)
```
rename to be clearer it is internal

SparkiDev commented on src/ssl.c:
```diff
@@ -48969,16 +51699,271 @@ int wolfSSL_X509_REQ_sign(WOLFSSL_X509 *req, WOLFSSL_EVP_PKEY *pkey,
+    }
+
+    /* return error if not built for libest */
+#ifdef HAVE_LIBEST
```
Move check up.

SparkiDev commented on wolfcrypt/src/asn.c:
```diff
@@ -8680,18 +8806,23 @@ static int DecodeCertExtensions(DecodedCert* cert)
+#ifdef WOLFSSL_CERT_REQ
+    if (!cert->isCSR)
+#endif
+    { /* Not included in CSR */
```
But Context-Specific 0 should be!

julek-wolfssl replied:
But a CRL doesn't use this API? Am I missing something here?

SparkiDev replied:
Certificate Request not a CRL.
See https://tools.ietf.org/html/rfc5280

SparkiDev replied:
Sorry. CSR not CRL.
Don't see extension in the RFC 2986 but do see attributes in a [0].

julek-wolfssl replied:
This part is omitted on purpose in a CSR. A CSR cannot have extensions like a cert would but it can have an Extension Request attribute. [In this commit I added a CSR with an Extension Request attribute.](https://github.com/wolfSSL/wolfssl/pull/3439/commits/fc3fa66d5afdbfe4b709dc27447b41d1eefde4a6) The structure of this CSR is (note the lack of a [0]):
```
julek@julek-MS-7A38:~/Projects/libest/test/UT$ openssl asn1parse -in csr.ext.der -inform DER -i
    0:d=0  hl=4 l= 562 cons: SEQUENCE          
    4:d=1  hl=4 l= 411 cons:  SEQUENCE          
    8:d=2  hl=2 l=   1 prim:   INTEGER           :03
   11:d=2  hl=2 l=  20 cons:   SEQUENCE          
   13:d=3  hl=2 l=  18 cons:    SET               
   15:d=4  hl=2 l=  16 cons:     SEQUENCE          
   17:d=5  hl=2 l=   3 prim:      OBJECT            :commonName
   22:d=5  hl=2 l=   9 prim:      UTF8STRING        :localhost
   33:d=2  hl=3 l= 159 cons:   SEQUENCE          
   36:d=3  hl=2 l=  13 cons:    SEQUENCE          
   38:d=4  hl=2 l=   9 prim:     OBJECT            :rsaEncryption
   49:d=4  hl=2 l=   0 prim:     NULL              
   51:d=3  hl=3 l= 141 prim:    BIT STRING        
  195:d=2  hl=3 l= 221 cons:   cont [ 0 ]        
  198:d=3  hl=2 l=  31 cons:    SEQUENCE          
  200:d=4  hl=2 l=   9 prim:     OBJECT            :challengePassword
  211:d=4  hl=2 l=  18 cons:     SET               
  213:d=5  hl=2 l=  16 prim:      PRINTABLESTRING   :IGCu/xNL4/0/wOgo
  231:d=3  hl=3 l= 185 cons:    SEQUENCE          
  234:d=4  hl=2 l=   9 prim:     OBJECT            :Extension Request
  245:d=4  hl=3 l= 171 cons:     SET               
  248:d=5  hl=3 l= 168 cons:      SEQUENCE          
  251:d=6  hl=3 l= 149 cons:       SEQUENCE          
  254:d=7  hl=2 l=   3 prim:        OBJECT            :X509v3 Subject Alternative Name
  259:d=7  hl=3 l= 141 prim:        OCTET STRING      [HEX DUMP]:30818A8704AC12340287047F0000018210626F626361742E636973636F2E636F6D82106A696D626F622E636973636F2E636F6D82136C6F63616C686F73742E636973636F2E636F6D8704AC12340287047F0000018210626F626361742E636973636F2E636F6D82106A696D626F622E636973636F2E636F6D82136C6F63616C686F73742E636973636F2E636F6D
  403:d=6  hl=2 l=  14 cons:       SEQUENCE          
  405:d=7  hl=2 l=   3 prim:        OBJECT            :X509v3 Key Usage
  410:d=7  hl=2 l=   1 prim:        BOOLEAN           :255
  413:d=7  hl=2 l=   4 prim:        OCTET STRING      [HEX DUMP]:030205E0
  419:d=1  hl=2 l=  13 cons:  SEQUENCE          
  421:d=2  hl=2 l=   9 prim:   OBJECT            :sha256WithRSAEncryption
  432:d=2  hl=2 l=   0 prim:   NULL              
  434:d=1  hl=3 l= 129 prim:  BIT STRING        
```
CSR attributes are read in `ParseCertRelative` where there is a comment `/* Read attributes */`.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -8253,18 +8374,23 @@ WOLFSSL_X509_EXTENSION* wolfSSL_X509_set_ext(WOLFSSL_X509* x509, int loc)
+    if (!x509->isCSR)
+#endif
+    {
+        if (input[idx++] != ASN_EXTENSIONS) {
```
This function only services WOLFSSL_X509 structs so I'm not sure why would we look for CRL characteristics?

julek-wolfssl replied:
For CSR, `cert.extensions` points directly to the extensions struct in the Extensions Request attribute (if it exists). This is what the `cert.extensions` struct points to:
```
julek@julek-MS-7A38:~/Projects/wolfssl-julek$ openssl asn1parse -in /tmp/ext.der -inform DER -i
    0:d=0  hl=3 l= 168 cons: SEQUENCE          
    3:d=1  hl=3 l= 149 cons:  SEQUENCE          
    6:d=2  hl=2 l=   3 prim:   OBJECT            :X509v3 Subject Alternative Name
   11:d=2  hl=3 l= 141 prim:   OCTET STRING      [HEX DUMP]:30818A8704AC12340287047F0000018210626F626361742E636973636F2E636F6D82106A696D626F622E636973636F2E636F6D82136C6F63616C686F73742E636973636F2E636F6D8704AC12340287047F0000018210626F626361742E636973636F2E636F6D82106A696D626F622E636973636F2E636F6D82136C6F63616C686F73742E636973636F2E636F6D
  155:d=1  hl=2 l=  14 cons:  SEQUENCE          
  157:d=2  hl=2 l=   3 prim:   OBJECT            :X509v3 Key Usage
  162:d=2  hl=2 l=   1 prim:   BOOLEAN           :255
  165:d=2  hl=2 l=   4 prim:   OCTET STRING      [HEX DUMP]:030205E0
```
Therefore the `ASN_EXTENSIONS` for CSR check can be skipped.

SparkiDev replied:
OK. The CSR parsing doesn't put the [0] into the extension data but the certificate parsing does put the [3].
Inconsistent but is what it is.

julek-wolfssl committed:
wolfSSL_BIO_do_connect should look for a socket bio in the chain

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15270,6 +15550,76 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+     * @param b BIO object
+     * @return WOLFSSL_SUCCESS on success and WOLFSSL_FAILURE on failure
+     */
+    long wolfSSL_BIO_do_connect(WOLFSSL_BIO *b)
```
I checked what other BIOs do and they just delegate the call to the next BIO in the chain. I changed it so that `wolfSSL_BIO_do_connect` will look for a BIO socket to perform this operation on. Good catch!

julek-wolfssl committed:
Sanity check protocol version.

julek-wolfssl committed:
Refactor extension stack generation

julek-wolfssl commented:
Known customer configs B — Failed  is fixed by #3564

julek-wolfssl commented on src/ssl.c:
```diff
@@ -18824,33 +19271,790 @@ void wolfSSL_GENERAL_NAME_free(WOLFSSL_GENERAL_NAME* name)
+        return 0;
+
+#ifndef NO_MD5
+    strLen = (int)XSTRLEN(str);
```
OpenSSL uses some custom hashing algorithm but comment to it is that it is "not as good as MD5" so I figured that MD5 will do fine for the purpose of this function. And the produced hashes don't need to line up between OpenSSL and wolfSSL. The hashes are for internal indexing only.

SparkiDev replied:
The comment above seems inconsistent then.

julek-wolfssl replied:
Changed comment to be more consistent.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -18824,33 +19271,790 @@ void wolfSSL_GENERAL_NAME_free(WOLFSSL_GENERAL_NAME* name)
+    return NULL;
+}
+
+WOLFSSL_CONF_VALUE *wolfSSL_CONF_get_section(WOLFSSL_CONF *conf,
```
Added `_CONF_new_section` and `_CONF_get_section`.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -20180,6 +21408,267 @@ void wolfSSL_sk_CIPHER_free(WOLF_STACK_OF(WOLFSSL_CIPHER)* sk)
+ * @param num Number of fields in each row.
+ * @return
+ */
+WOLFSSL_TXT_DB *wolfSSL_TXT_DB_read(WOLFSSL_BIO *in, int num)
```
Well there is a test in `api.c` that tests loading `./tests/TXT_DB.txt` using this API. Jacob ran it through a fuzzing tool that had some creative inputs that helped find some additional errors.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -21713,10 +23222,78 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
+    }
+
+#ifndef NO_WOLFSSL_STUB
+    void wolfSSL_X509_get0_signature(const WOLFSSL_ASN1_BIT_STRING **psig,
```
It is used in the libest unit tests to print information but the output is not used so a stub that doesn't do anything is enough to get the tests passing.

julek-wolfssl committed:
Code review names changes and refactoring

kaleb-himes reviewed (changes requested):
Code in ssl.c (outside the FIPS module boundary) calls new function `wc_DhKeyCopy` without a FIPS macro guard.

`wc_DhKeyCopy` is implemented in dh.c (inside the FIPS module boundary) and will not be available in FIPS mode for cert 3389.

kaleb-himes commented on src/ssl.c:
```diff
@@ -29413,6 +31274,43 @@ int SetDhInternal(WOLFSSL_DH* dh)
+        return NULL;
+    }
+
+    if (wc_DhKeyCopy((DhKey*)dh->internal, (DhKey*)ret->internal) != MP_OKAY) {
```
Code in ssl.c (outside the FIPS module boundary) calls new function wc_DhKeyCopy without a FIPS macro guard.

wc_DhKeyCopy is implemented in dh.c (inside the FIPS module boundary) and will not be available in FIPS mode for cert 3389.

kaleb-himes commented on wolfcrypt/src/dh.c:
```diff
@@ -2079,6 +2079,45 @@ int wc_DhAgree(DhKey* key, byte* agree, word32* agreeSz, const byte* priv,
 }
 
 #ifdef WOLFSSL_DH_EXTRA
+WOLFSSL_LOCAL int wc_DhKeyCopy(DhKey* src, DhKey* dst)
```
Won't be available in FIPSv2 (cert 3389)

julek-wolfssl committed:
Add CSR test with Extension Request attribute

julek-wolfssl committed:
Call X509_REQ_get_extensions and X509_get_ext_by_NID on a CSR object

julek-wolfssl commented on src/ssl.c:
```diff
@@ -9147,16 +9388,21 @@ int wolfSSL_X509_get_ext_by_NID(const WOLFSSL_X509* x509, int nid, int lastPos)
+    if (!x509->isCSR)
+#endif
+    {
+        if (input[idx++] != ASN_EXTENSIONS) {
```
As above, for CSR, cert.extensions points directly to the extensions struct in the Extensions Request attribute (if it exists).

julek-wolfssl committed:
Refactor wc_CheckPrivateKey

- Change wc_CheckPrivateKey to wc_CheckPrivateKeyCert and wc_CheckPrivateKey
- wolfSSL_X509_check_private_key no longer needs to decode cert to check key
- Fix scope in api.c

julek-wolfssl committed:
Fix after rebase

julek-wolfssl committed:
Use CSR with smaller key size 4096 -> 2048
